### DISCUSSION TITLE
Custom two factor authenticator, works but does not appear in Account Console

### DISCUSSION THREAD

**Discussion by korridor** [keycloak/keycloak#21066] [Q&A]
I started to use the following plugin for a custom SMS-based second factor: [https://github.com/cooperlyt/keycloak-phone-provider](https://github.com/cooperlyt/keycloak-phone-provider)

This plugin seams to work great.
I can add the SMS verification as a required action.
The only thing that is missing, is that I can not configure it in the Account Console.
See screenshot:

<img width="1557" alt="image" src="https://github.com/keycloak/keycloak/assets/26689068/ad7118b6-ce5b-42ba-a078-ba19697d9e0e">

I cloned the project and looked into the code, but I can not find the issue.
I compared it to the example provided by Keycloak in the `example` folder, and I can not see any difference that explains why it should work there but not in the SMS provider.
Link to the Keycloak example: [https://github.com/keycloak/keycloak/tree/main/examples/providers/authenticator](https://github.com/keycloak/keycloak/tree/main/examples/providers/authenticator)
If I understand Keycloak code correctly, the `getCredentialTypeMetadata` looks to me like the place where this should be configured, but this looks fine to me in the SMS provider.

Keycloak example: https://github.com/keycloak/keycloak/blob/0ca9b217654e5ee69ebd16657a36630b2a4c642a/examples/providers/authenticator/src/main/java/org/keycloak/examples/authenticator/SecretQuestionCredentialProvider.java#L94
SMS provider: https://github.com/cooperlyt/keycloak-phone-provider/blob/a7acd58101e79dfaa048261470ab4cba3fe16249/keycloak-phone-provider/src/main/java/cc/coopersoft/keycloak/phone/credential/PhoneOtpCredentialProvider.java#L128

Any tips regarding what could be the problem, or what I should check, are highly appreciated.

**Comment by teunis90** [comment]
@korridor I was playing with the same thing today, and I figured out that this is because of a missing interface: `CredentialInputUpdater`. It will require the implementation of some methods as well.

```
public class PhoneOtpCredentialProvider implements CredentialProvider<PhoneOtpCredentialModel>, CredentialInputValidator {
->
public class PhoneOtpCredentialProvider implements CredentialProvider<PhoneOtpCredentialModel>, CredentialInputValidator, CredentialInputUpdater {
```

Hope this will help someone.

**Reply by korridor** [reply]
Thank you so much for the input! Did you make it work after adding the interface and if yes, would it be possible for you to share the implementations of the functions required by the interface?

**Reply by teunis90** [reply]
I was wrong; I found the real issue after spending another day on it. 

The reason why I did not pop up for me was because:

1. Required Action `getReferenceCategory()` should have the same value as `PhoneOtpCredentialProvider.java.getId()`, which in turn is defined in `PhoneOtpCredentialModel.java `.

https://github.com/cooperlyt/keycloak-phone-provider/blob/83fce1d60b259def154838112aec8067a9d7cd16/keycloak-phone-provider/src/main/java/cc/coopersoft/keycloak/phone/authentication/authenticators/browser/SmsOtpMfaAuthenticatorFactory.java#L30-L33

https://github.com/cooperlyt/keycloak-phone-provider/blob/83fce1d60b259def154838112aec8067a9d7cd16/keycloak-phone-provider/src/main/java/cc/coopersoft/keycloak/phone/credential/PhoneOtpCredentialProvider.java#L103-L106

https://github.com/cooperlyt/keycloak-phone-provider/blob/83fce1d60b259def154838112aec8067a9d7cd16/keycloak-phone-provider/src/main/java/cc/coopersoft/keycloak/phone/credential/PhoneOtpCredentialModel.java#L20

In this example, the change should be in `PhoneOtpCredentialModel.java ` line 20:

```
    @Override
    public String getReferenceCategory() {
-        return "OTP over SMS";
+        return "phone-otp";
    }
```

2. It only pops ups when the authenticator is configured in the browser flow.

I figured it out by following the `/credentials` endpoint with a step debugger. See line 243'ish. This is where the `getReferenceCategory` value gets compared.

https://github.com/keycloak/keycloak/blob/92cc7708fe79945a30d1f75ea338d0780cc2cef6/services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java#L150-L268

See line 

Just a remark: I used Coopersoft's `keycloak-phone-provider.git` as a start for my project. Unfortunately, I can't share my code at this point. I hope this will resolve your issue.

**Reply by korridor** [reply]
Thanks so much for the input.
 
I updated the following line to `phone-otp` like this:

https://github.com/cooperlyt/keycloak-phone-provider/blob/83fce1d60b259def154838112aec8067a9d7cd16/keycloak-phone-provider/src/main/java/cc/coopersoft/keycloak/phone/authentication/authenticators/browser/SmsOtpMfaAuthenticatorFactory.java#L30-L33

```java
// in cc.coopersoft.keycloak.phone.authentication.authenticators.browser.SmsOtpMfaAuthenticatorFactory

    @Override
    public String getReferenceCategory() {
        return "phone-otp";
    }
```

I also activated SMS OTP and added it to a custom browser flow, but it still does not work for me.

<img width="1585" alt="image" src="https://github.com/keycloak/keycloak/assets/26689068/b798d4db-df9f-40b2-b0d6-325d25b8fe1c">
<img width="1566" alt="image" src="https://github.com/keycloak/keycloak/assets/26689068/5e50a66e-4e32-4741-a434-d66ae59e64f4">
<img width="1575" alt="image" src="https://github.com/keycloak/keycloak/assets/26689068/01561f9c-3479-4874-be89-cf317e7124fe">

Did you change anything else? Did you also implement the interface you were talking about previously?
